providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/model/FlagParser.java

```diff
@@ -0,0 +1,131 @@
+    private static final String REPLACER_FORMAT = "\"\\$ref\":(\\s)*\"%s\"";
+    private static final String SCHEMA_RESOURCE = "json-schema-v0.1.1";
+
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    private static final Map<String, Pattern> PATTERN_MAP = new HashMap<>();
+    private static final Pattern REG_BRACKETS = Pattern.compile("^[^{]*\\{|}[^}]*$");
+
+    private static JsonSchema SCHEMA_VALIDATOR;
+
+    static {
+        try {
+            final URL url = FlagParser.class.getClassLoader().getResource(SCHEMA_RESOURCE);
+            if (url == null) {
+                log.log(Level.WARNING, String.format("Resource %s not found", SCHEMA_RESOURCE));
+            } else {
+                byte[] bytes = Files.readAllBytes(Paths.get(url.getPath()));
+                String schemaString = new String(bytes, StandardCharsets.UTF_8);
+                JsonSchemaFactory instance = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V7);
+                SCHEMA_VALIDATOR = instance.getSchema(schemaString);
+            }
+        } catch (Throwable e) {
+            // log only, do not throw
+            log.log(Level.WARNING,
+                    String.format("Error loading resource %s, schema validation will be skipped", SCHEMA_RESOURCE), e);
+        }
+    }
+
+    /**
+     * Parse {@link String} for feature flags.
+     */
+    public static Map<String, FeatureFlag> parseString(final String configuration) throws IOException {
+        if (SCHEMA_VALIDATOR != null) {
+            try (JsonParser parser = MAPPER.createParser(configuration)) {
+                Set<ValidationMessage> validationMessages = SCHEMA_VALIDATOR.validate(parser.readValueAsTree());
+
+                if (!validationMessages.isEmpty()) {
+                    throw new IllegalArgumentException(
+                            String.format("Failed to parse configurations. %d validation " + "error(s) reported.",
+                                    validationMessages.size()));
+                }
+            }
+        }
+
+        final String transposedConfiguration = transposeEvaluators(configuration);
+
+        final Map<String, FeatureFlag> flagMap = new HashMap<>();
+
+        try (JsonParser parser = MAPPER.createParser(transposedConfiguration)) {
+            final TreeNode treeNode = parser.readValueAsTree();
+            final TreeNode flagNode = treeNode.get(FLAG_KEY);
+
+            if (flagNode == null) {
+                throw new IllegalArgumentException("No flag configurations found in the payload");
+            }
+
+            final Iterator<String> it = flagNode.fieldNames();
+            while (it.hasNext()) {
+                final String key = it.next();
+                flagMap.put(key, MAPPER.treeToValue(flagNode.get(key), FeatureFlag.class));
+            }
+        }
+
+        return flagMap;
+    }
+
+    private static String transposeEvaluators(final String configuration) throws IOException {
```

Comment:
This reminds me that I should add shared evaluators to the e2e test harness: https://github.com/open-feature/flagd-testbed/pull/54. I didn't add those.

Reply:
Yup, we need $ref rules and validate them in e2e tests :)